### DISCUSSION TITLE
Help needed please

### DISCUSSION THREAD

**Discussion by JohnnyAstroman** [cinderblocks/libremetaverse#88] [General]
I want to learn how libremetaverse works and eventually contribute. I just downloaded the zip file and the solution does not appear to build, i am getting lots of missing assemblies. What do  i need to install and reference please ?. The old libmetaverse did not have this issue.

Severity	Code	Description	Project	File	Line	Suppression State
Error	CS1069	The type name 'Bitmap' could not be found in the namespace 'System.Drawing'. This type has been forwarded to assembly 'System.Drawing.Common, Version=0.0.0.0,

Am i correct in that you have to use NuGet to install some package ?. Sorry never usesd NuGet before

I am using Install Visual Studio 2022 (really for the first time) so i think you can add packages here. What steps do i need to take
please?

looking in the nuget package manager menu option i get "System.Drawing.common" - not available in this source "
is this important ? i see installed version 7 next to it, so i am puzzled

**Comment by JohnnyAstroman** [comment]
so i have fixed this and sucessfully built the solution. I know i am a newbee but i want to document my leanring process for others  to not have the same issues i had.
https://www.youtube.com/watch?v=bYKymAixguA 
i added  the NuGet package as in the vdeo and it allowed me to build the solution , now testing it and it seems to work, must less spam in the console window compared to the old libopenmetaverse version.

**Comment by JohnnyAstroman** [comment]
Now trying to build the solution on a raspberry PI, and have a new issue when trying to run the exe which did seem to sucessfully build,

I ran the following :-
dotnet restore LibreMetaverse.ReleaseNoGui.sln               seemed to work ok 
dotnet msbuild LibreMetaverse.ReleaseNoGui.sln             seemed to work ok , i got an exe and dll.
dotnet --version gives me 
7.0.400

dotnet --list-runtimes
Microsoft.AspNetCore.App 6.0.21 [/home/pi/.dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 7.0.10 [/home/pi/.dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.NETCore.App 6.0.21 [/home/pi/.dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 7.0.10 [/home/pi/.dotnet/shared/Microsoft.NETCore.App]

dotnet TestClient.dll
Unhandled exception. System.IO.FileLoadException: Could not load file or assembly 'TestClient, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null'.
Aborted

 ./TestClient
Unhandled exception. System.IO.FileLoadException: Could not load file or assembly 'TestClient, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null'.
Aborted

**Comment by Misterblue** [comment]
For the TestClient errors, are you in the directory that contains the TestClient dll?

For learning how to use libreMetaverse, exploring TestClient is your best starting point. It has the code that logs in an account and then has examples of many of the possible operations in TestClient/Commands/{Topic}/{Command}.cs.

**Reply by JohnnyAstroman** [reply]
Hi Misterblue, yes i am in the directory that contains the TesClient dll. I am currently trying on  another raspberry pi that only has .NET 7 on. On the "above" raspberry pi I first of all installed 6 and then 7 , and them removed version 6. Possibly i have broken dotnet but  the build worked, so i think it is something more fundamental than that.

Tried on another  "pi" and it failed the same way with a purely clean .NET 7 installed on. Both "pi" also had "mono" installed as well

i am now settign up a new Raspberry pi with the latest "bullseye" images both 32 and 64 bit, and will see what happens with these. This is Debian Linux version 11

I will update here on the weekend once i have the test Pi ready to go

**Comment by JohnnyAstroman** [comment]
Files in the run timelibrary  after 


-rwxrwxrwx 1 pi pi 201312 Aug 19 13:41 TestClient
-rwxrwxrwx 1 pi pi  47566 Aug 19 13:41 TestClient.deps.json
-rwxrwxrwx 1 pi pi 192512 Aug 19 13:41 TestClient.dll
-rwxrwxrwx 1 pi pi  77760 Aug 19 13:41 TestClient.pdb
-rwxrwxrwx 1 pi pi    139 Aug 19 13:41 TestClient.runtimeconfig.json


pi@xxx:/media/HDD1/libremetaverse/libremetaverse-master/bin/net7.0 $ ./TestClient
Unhandled exception. System.IO.FileLoadException: Could not load file or assembly 'TestClient, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null'.
Aborted

this does not work on a fresh installed Rapsberry pi with 64bit Debian. 
pi@xxx:/media/HDD1/libremetaverse/libremetaverse-master/bin/net7.0 $ dotnet --version
7.0.400

what have i done wrong ?

dotnetr was installed as per https://learn.microsoft.com/en-us/dotnet/iot/deployment
curl -sSL https://dot.net/v1/dotnet-install.sh | bash /dev/stdin --channel STS
echo 'export DOTNET_ROOT=$HOME/.dotnet' >> ~/.bashrc
echo 'export PATH=$PATH:$HOME/.dotnet' >> ~/.bashrc
source ~/.bashrc

dotnet restore and 
dotnet build appeared to complete sucessfully.,

**Reply by cinderblocks** [reply]
I don't believe that I have build ARM compatible libs for decoding J2K. We're kind of at a crossroads here since `Bitmap` was deprecated, and my OpenJPEG shim makes heavy use of it. I will see if I can either come up with a no-image LibreMetaverse or get ARM native libraries built.

**Reply by JohnnyAstroman** [reply]
thank you cinderblocks, i do intend to repay the work and learn so that i can contribute. I am learning C# but started off programming in Basic and Pascal.  I have visual studio 2022 Enterprise edition,  but my key use case is to devlop in windows and use on Raspberry PI's.
The old OpenMetaversse used to work fine with MONO on a Raspberry Pi.
I am spending  time stepping through TestClient at the moment, and would love to eventually create some class diagrams, and write some documentation. I would love to see TestClient do the auto-relog when a sim restarts as well. I could however help to test if there was any value in that.

**Reply by JohnnyAstroman** [reply]
since i really just want a linux version to log a bot into second life, would it be easier to just create a small app that logged in and kept a bot logged on. Ideally i want it to relog after a sim restart. There must be an example piece of code that does this somewhere.

**Comment by JohnnyAstroman** [comment]
Finally got back to this and afer some self stufy i have the solution built in both Linux and windows. Now i can start to study the code in Q1,Q2 2025
